mmengine/_strategy/deepspeed.py

```diff
@@ -517,35 +521,56 @@ def save_checkpoint(
             extra_ckpt (dict, optional): Extra checkpoint to save.
                 Defaults to None.
         """
-        if extra_ckpt is None:
-            extra_ckpt = dict()
-        if 'meta' not in extra_ckpt:
-            extra_ckpt['meta'] = dict()
-        extra_ckpt['meta'].update(
-            seed=self.seed,
-            time=time.strftime('%Y%m%d_%H%M%S', time.localtime()),
-            mmengine=mmengine.__version__ + get_git_hash(),
-        )
-
-        if save_optimizer and hasattr(self, 'optim_wrapper'):
-            # The key can not be 'optimizer', otherwise error will be thrown
-            # when loading or resuming checkpoint.
-            extra_ckpt['optim_wrapper'] = self.optim_state_dict()
-
-        if save_param_scheduler and hasattr(self, 'param_schedulers'):
-            extra_ckpt['param_schedulers'] = self.scheduler_state_dict()
+        if (not save_optimizer
+                and self.config.zero_optimization.get('stage', 0) == 3
+                and not self.config.zero_optimization.get(
+                    'stage3_gather_16bit_weights_on_model_save', False)):
+            print_log(
+                'Configured to `save_optimizer=False`, but currently using '
+                "DeepSpeed's ZeRO stage 3 with "
+                '`stage3_gather_16bit_weights_on_model_save=False`. In '
+                'this configuration, the model cannot be saved properly '
+                'and will be saved with the optimizer state. '
+                'To support `save_optimizer=False`, please set '
+                '`stage3_gather_16bit_weights_on_model_save=True` in your '
+                'DeepSpeed config.',
+                logger='current',
+                level=logging.WARNING)
+            save_optimizer = True
 
         dirname, basename = osp.split(filename)
-        if digit_version(deepspeed.__version__) >= digit_version('0.10.1'):
-            self.model.save_checkpoint(
-                dirname,
-                tag=basename,
-                client_state=extra_ckpt,
-                save_latest=False,
-                exclude_frozen_parameters=self.exclude_frozen_parameters)
+        if not save_optimizer:
+            self.model.save_16bit_model(
+                save_dir=dirname, save_filename=basename)
         else:
-            self.model.save_checkpoint(
-                dirname,
-                tag=basename,
-                client_state=extra_ckpt,
-                save_latest=False)
+            if extra_ckpt is None:
+                extra_ckpt = dict()
+            if 'meta' not in extra_ckpt:
+                extra_ckpt['meta'] = dict()
+            extra_ckpt['meta'].update(
+                seed=self.seed,
+                time=time.strftime('%Y%m%d_%H%M%S', time.localtime()),
+                mmengine=mmengine.__version__ + get_git_hash(),
+            )
```

Comment:
Should extra_ckpt be saved when save_optimizer is False?

Reply:
`{'meta': meta, ..., 'state_dict': state_dict()}`
This is the standard structure for saved checkpoint (e.g., epoch_1.pth).

This PR saves checkpoint by `self.model.save_16bit_model` when `save_optimzer==False`, and this method cannot involve the `extra_ckpt` to saved checkpoint because there is no such interface.

To save `extra_ckpt`, one way is to copy-paste the related code to get `state_dict()` first and construct a standard checkpoint
https://github.com/microsoft/DeepSpeed/blob/b81bed69a8db3c1e3263c27f48dcecf12b354931/deepspeed/runtime/engine.py#L3570-L3580

What do you think of this?

Reply:
It would be better to follow the standard structure `{'meta': meta, ..., 'state_dict': state_dict()}`.